File: python/src/tradingflow/operators/metrics/turnover.py
"""Portfolio turnover metric."""

from dataclasses import dataclass, field

import numpy as np

from ...views import ArrayView
from ...operator import Operator, Notify
from ...types import Array, Handle, NodeKind


@dataclass(slots=True)
class TurnoverState:
    num_stocks: int
    prev: np.ndarray = field(default_factory=lambda: np.empty(0))
    initialized: bool = False


class Turnover(
    Operator[
        tuple[Handle[Array[np.float64]]],
        Handle[Array[np.float64]],
        TurnoverState,
    ]
):
    """Per-rebalance portfolio turnover.

    On every update of the input weight vector, emits the L1 norm of
    the change since the previous update:

        turnover_t = sum_i |w_{t,i} - w_{t-1,i}|

    For long-only portfolios that sum to 1, turnover lies in ``[0, 2]``:
    ``0`` means no change, ``2`` means a complete liquidation and
    re-investment.

    The first update is a warmup: the operator caches the weights and
    emits ``NaN``.  All subsequent updates emit a finite turnover value.

    NaN handling: positions ``w_{t,i}`` and ``w_{t-1,i}`` that are
    non-finite are treated as ``0`` before computing the difference,
    so a stock going from active to missing (or vice versa) contributes
    its full weight to the turnover.

    Output is a scalar. ``Record(output)`` produces a plottable time
    series; downstream ``RollingMean`` yields an average turnover.

    Parameters
    ----------
    weights
        Soft position weights, shape ``(N,)``.  Typically the output
        of a portfolio constructor.
    """

    def __init__(self, weights: Handle) -> None:
        assert len(weights.shape) == 1
        self._num_stocks = weights.shape[0]

        super().__init__(
            inputs=(weights,),
            kind=NodeKind.ARRAY,
            dtype=np.float64,
            shape=(),
            name=type(self).__name__,
        )

    def init(self, inputs: tuple, timestamp: int) -> TurnoverState:
        return TurnoverState(num_stocks=self._num_stocks)

    @staticmethod
    def compute(
        state: TurnoverState,
        inputs: tuple[ArrayView[np.float64]],
        output: ArrayView[np.float64],
        timestamp: int,
        notify: Notify,
    ) -> bool:
        current = np.where(np.isfinite(inputs[0].value()), inputs[0].value(), 0.0)

        if not state.initialized:
            state.prev = current
            state.initialized = True
            return False

        turnover = float(np.sum(np.abs(current - state.prev)))
        state.prev = current
        output.write(np.array(turnover, dtype=np.float64))
        return True

File: crates/tradingflow/src/operators/metric/turnover.rs
use num_traits::Float;
use std::marker::PhantomData;
use tradingflow_data::Layout;

use crate::data::{Array, ArrayView, Instant, Scalar};
use crate::graph::Segment;
use crate::ports::ArrayPort;

/// Operator signature for [`turnover`].
pub struct Turnover<T: Scalar + Float> {
    _marker: PhantomData<fn() -> T>,
}

impl<T: Scalar + Float> Turnover<T> {
    pub fn new() -> Self {
        Self {
            _marker: PhantomData,
        }
    }
}

impl<T: Scalar + Float> Default for Turnover<T> {
    fn default() -> Self {
        Self::new()
    }
}

/// Runtime state for [`Turnover`].
pub struct TurnoverState<T: Scalar + Float> {
    prev: Vec<T>,
    out: Array<T, 0>,
}

impl<T: Scalar + Float> Segment for Turnover<T> {
    type Inputs = ArrayPort<T, 1>;
    type Outputs = ArrayPort<T, 0>;
    type Context = Instant;
    type State = TurnoverState<T>;

    fn init(self, weights: ArrayView<'_, T, 1>) -> Self::State {
        let len = weights.layout().len();
        TurnoverState {
            prev: vec![T::zero(); len],
            out: Array::scalar(T::zero()),
        }
    }

    fn reset<'a, 'b: 'a>(
        _: ArrayView<'a, T, 1>,
        state: &'b mut Self::State,
    ) -> ArrayView<'a, T, 0> {
        state.out.view()
    }

    fn compute<'a, 'b: 'a>(
        weights: ArrayView<'a, T, 1>,
        state: &'b mut Self::State,
        _: &Self::Context,
    ) -> ArrayView<'a, T, 0> {
        assert!(
            weights.layout().len() == state.prev.len(),
            "turnover: input length mismatch"
        );
        for w in weights.iter() {
            assert!(w.is_finite(), "turnover: input values must be finite");
        }
        crate::data::array::for_each(weights, |j, &w| {
            *state.out = *state.out + (w - state.prev[j]).abs();
            state.prev[j] = w;
        });
        state.out.view()
    }
}

/// Cumulative turnover: the sum of L1 norm of changes in a weight vector.
pub fn turnover<T: Scalar + Float>()
-> impl Segment<Inputs = ArrayPort<T, 1>, Outputs = ArrayPort<T, 0>, Context = Instant> {
    Turnover::new()
}
